teps/0058-graceful-pipeline-run-termination.md

```diff
@@ -136,32 +149,59 @@ Consider including folks that also work outside the WGs or subproject.
 
 ### User Experience (optional)
 
-<!--
-Consideration about the user experience. Depending on the area of change,
-users may be task and pipeline editors, they may trigger task and pipeline
-runs or they may be responsible for monitoring the execution of runs,
-via CLI, dashboard or a monitoring system.
-
-Consider including folks that also work on CLI and dashboard.
--->
+Support for additional command or flags in `tkn` CLI should be considered.
 
 ### Performance (optional)
 
 No impact on performance.
 
 ## Design Details
 
-<!--
-This section should contain enough information that the specifics of your
-change are understandable.  This may include API specs (though not always
-required) or even code snippets.  If there's any ambiguity about HOW your
-proposal will be implemented, this is the place to discuss them.
-
-If it's helpful to include workflow diagrams or any other related images,
-add them under "/teps/images/". It's upto the TEP author to choose the name
-of the file, but general guidance is to include at least TEP number in the
-file name, for example, "/teps/images/NNNN-workflow.jpg".
--->
+In this proposal we need to consider following cases:
+
+1. User sets `spec.status` to "PipelineRunStopped" in `PipelineRun` with running tasks, but `finally` section is empty.
+
+    In this case a graceful termination would ack the same as a pipeline run cancellation.
+    
+    - `spec.status` in all running `TaskRun` is patched to "TaskRunCancelled"
+    - `spec.status` in all running `Run` is patched to "RunCancelled"
+    - `PipelineRun` condition after graceful termination is:
+      ```yaml
+      { type: Succeeded, Status: False, Reason: Cancelled } 
+      ```
+
+2. User sets `spec.status` to "PipelineRunStopped" in `PipelineRun` with running tasks and non-empty `finally` section.
+
+    In this case a graceful termination cancels all running task runs and waits for final tasks to be processed.
+    
+    - `spec.status` in all running `TaskRun` is patched to "TaskRunCancelled"
+    - `spec.status` in all running `Run` is patched to "RunCancelled"
+    - `PipelineRun` condition after graceful termination is:
+      ```yaml
+      { type: Succeeded, Status: Unknown, Reason: PipelineRunStopping } 
+      ```    
```

Comment:
Again, I would like to know the `pipelineRun` state transitioning in this case:

`Running` -> `PipelineRunStopping` (while `finally` tasks are executing) -> `Cancelled`?

It makes me think the state label `PipelineRunStopping` might not fit well with the graceful termination. The `pipelineRun` is actually executing `finally` tasks and might run for the considerable amount of time when `pipelineRun` is in `stopping` state.

Please feel free to ignore me if others don't see any issues with it. I was thinking of renaming `PipelineRunStopping` to `PipelineRunExitHandler` or `PipelineRunGraciouslyCancelled` 🤔 , not sure if its any better 😜

Reply:
@pritidesai please see my previous comment.

Reply:
I came up with one more name `PipelineRunFinally` 😉

Reply:
`PipelineRunCancelledReallyGracefully` :D